### PR TITLE
New Resource: aws s3 bucket

### PR DESCRIPTION
**Pull request by dromazmj**
* Adds functionality for checking the permissions of specific users in a buckets
    - can check the owner, authorized users, everyone and log group
* Adds functionality for checking if the bucket has any public files

### CODE REVIEW

File: docs/resources/aws_s3_bucket.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+<br>
+
+## Examples
+
+The following examples show how to use this InSpec audit resource.
+
+### Test a buckets permissions
+
+    describe aws_s3_bucket(name: 'test_bucket') do
+      its('permissions_owner') { should cmp ['FULL_CONTROL'] }
```
Comment (aaronlippold) [inline docs/resources/aws_s3_bucket.md]: Left over from last push

File: docs/resources/aws_s3_bucket.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      # Check examples of 'authUsers'
+      its('permissions.authUsers') { should be_in ['READ'] }
+
+      # Check examples of 'everyone'
+      its('permissions.everyone') { should be_in [] }
+
+      # Check examples of the 'logGroup'
+      its('permissions.logGroup') { should be_in ['WRITE'] }
+    end
+
```
Comment (aaronlippold) [inline docs/resources/aws_s3_bucket.md]: Should we have a test in here for the region?

File: test/unit/resources/aws_s3_bucket_test.rb [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  def setup
+    AwsS3Bucket::BackendFactory.select(AwsMSBSB::Basic)
+  end
+
+  def test_property_name
+    assert_equal('Public Bucket', AwsS3Bucket.new('Public Bucket').name)
+  end
+
+  #-----------------------------------------------------#
+  # Testing Propertys of a public bucket
```
Comment (clintoncwolfe) [inline test/unit/resources/aws_s3_bucket_test.rb]: Properties, not Propertys :)

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
My overall impression is that is has a lot of good work in it.  This is really two PRs: create the resource, and then add ACL-based access analysis.  I'd want to see something in the docs about limitations (for example, scalability issues on object ACLs, and no bucket profile analysis).  NOT asking for those to be added to this PR - smaller, more focused PRs are more desirable.

I haven't run the tests yet on this; once the fixes are in, I'll move on to that.  Great contribution, and thanks @dromazmj !

File: docs/resources/aws_s3_bucket.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    describe aws_s3_bucket(name: 'test_bucket') do
+      it { should_not have_public_files }
+    end
+
+<br>
+
+## Supported Properties
+
+### permissions (Hash)
```
[the remark this answers is not in the document]
Comment (aaronlippold) [reply docs/resources/aws_s3_bucket.md]: I agree for ongoing PRs ... are these ok for the first round or do you really want to split out these into two PRs each?

File: docs/resources/aws_s3_bucket.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+The `permissions` hash property is used for matching the permissions of specific users.
+
+    describe aws_s3_bucket('test_bucket') do
+      # Check what extension categories we have
```
[the remark this answers is not in the document]
Comment (aaronlippold) [reply docs/resources/aws_s3_bucket.md]: we do that below correct?

[the remark this answers is not in the document]
Comment (aaronlippold) [reply docs/resources/aws_s3_bucket.md]: Do you want us to remove that example? I think @dromazmj was just trying to show you could look at the whole set of permissions.

File: docs/resources/aws_s3_bucket.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply docs/resources/aws_s3_bucket.md]: I'm OK either way.  I'm just trying to let you know what affects the velocity of the review process.

File: docs/resources/aws_s3_bucket.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply docs/resources/aws_s3_bucket.md]: I'm saying the first set of examples are potentially confusing, while the remainder are excellent.  So like this:
```
describe aws_s3_bucket('test_bucket') do
    # Check examples of 'owner'
    its('permissions.owner') { should be_in ['FULL_CONTROL'] }

    # Check examples of 'authUsers'
    its('permissions.authUsers') { should be_in ['READ'] }

    # Check examples of 'everyone'
    its('permissions.everyone') { should be_in [] }

    # Check examples of the 'logGroup'
   its('permissions.logGroup') { should be_in ['WRITE'] }
end
```

Might be a better (clearer) matcher than `be_in` (perhaps `cmp`), the main point is the latter examples are good.

File: docs/resources/aws_s3_bucket.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      # Check examples of 'everyone'
+      its('permissions.everyone') { should be_in [] }
+
+      # Check examples of the 'logGroup'
+      its('permissions.logGroup') { should be_in ['WRITE'] }
+    end
+
+### objects (Hash)
+
+The `objects` hash property is used for testing the objects in a bucket.
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply docs/resources/aws_s3_bucket.md]: We generally lean toward fluent reading - we assume the profile should be understandable by an auditor, so `public_objects` is better to me than `objects.public`.

File: docs/resources/aws_s3_bucket.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      its('objects.public') { should eq [] }
+    end
+
+## Matchers
+
+This InSpec audit resource has the following special matchers. For a full list of available matchers (such as `exist`) please visit our [matchers page](https://www.inspec.io/docs/reference/matchers/).
+
+### have_public_files (alias: has_public_files)
+
+The `have_public_files` matcher tests if the S3 Bucket has any files that are open to the public. Returns a true if one or more objects in the bucket are public.  If no objects are public returns false.  Please visit https://blog.rackspace.com/3-things-aws-s3-security-stay-headlines for more details on what is public.
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply docs/resources/aws_s3_bucket.md]: Yes, I absolutely agree we should document what `have_public_objects` means, from a technical perspective.

File: test/integration/default/build/s3.tf [hunk 8]
```diff
@@ -0,0 +1,58 @@
+resource "random_id" "bucket_id" {
+  byte_length = 8
+}
+
+resource "aws_s3_bucket" "aws_demo_bucket" {
+  bucket        = "aws-demo-s3-bucket-${random_id.bucket_id.hex}"
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply test/integration/default/build/s3.tf]: The intent behind terraform.env (which is the workspace name, or `ENV['INSPEC_TERRAFORM_ENV']` from the Rakefile) is that it be somewhat unique but predicatable; I use cwolfe-01, cwolfe-02, etc.

**Pull request closed without merging**